### PR TITLE
Mass Codebase Refactor (with a Focus on Combat Systems)

### PR DESCRIPTION
**Pull request by gsemaj**
Big ol' codebase cleanup:
- New entity hierarchy to add support for non-mob combatants
- NPC Appearance Data is no longer stored, instead created when needed
- Less magic numbers and more utility structs
- Reworked Combat, AI, and Ability systems to be more granular & generalized to support ICombatants and (TODO) event hooks
- Source and header file (re)organization

### CODE REVIEW

**Comment by gsemaj** [comment]
okay, I think we got the big stuff out of the way. let the reviewing begin

**Review by CPunch — APPROVED** [review]
LGTM 😋 

as stated in another off-site discussion, I'd like to switch all of the memset() calls to just zero initializers where appropriate after merging, eg. instead of zeroing out with memset, just use ` = {0}` when declaring the array/struct

**Comment by gsemaj** [comment]
> LGTM 😋 
> 
> 
> 
> as stated in another off-site discussion, I'd like to switch all of the memset() calls to just zero initializers where appropriate after merging, eg. instead of zeroing out with memset, just use ` = {0}` when declaring the array/struct. I've gone ahead and opened an issue at https://github.com/OpenFusionProject/OpenFusion/issues/260

we'll go ahead and make that change in the formatting pass, as discussed offline

File: src/Chunking.cpp [hunk 1]
```diff
@@ -75,65 +75,65 @@ void Chunking::untrackEntity(ChunkPos chunkPos, const EntityRef& ref) {
         deleteChunk(chunkPos);
 }
 
-void Chunking::addEntityToChunks(std::set<Chunk*> chnks, const EntityRef& ref) {
+void Chunking::addEntityToChunks(std::set<Chunk*> chnks, const EntityRef ref) {
     Entity *ent = ref.getEntity();
-    bool alive = ent->isAlive();
+    bool alive = ent->isExtant();
```
[the remark this answers is not in the document]
Comment (dongresource) [reply src/Chunking.cpp:80]: I grepped the current version of the branch for all occurrences of `isAlive()`, and they all looked legit. Given that this branch is currently mostly up-to-date with `master`, that's that.

**Review by dongresource — APPROVED** [review]
Wrote down a bunch of nitpicks. Nothing left that really needs to be fixed before public testing. We can clean up some of the stuff I highlighted after merging; some of it is for the formatting pass; and some of it can just be left alone, probably.

I looked through each and every file, but obviously can't thoroughly review this much code. Mostly focused on things other than the combat and ability logic that we already thoroughly test in-game.

**Pull request merged**